### DISCUSSION TITLE
TLS/HTTPS Mock Server

### DISCUSSION THREAD

**Discussion by rcurran1221** [pactumjs/pactum#148] [General]
Hello All!

How do I enable HTTPS/TLS for a Pactum Mock Server? I can't seem to find any reference to it in the documentation.

Thanks!

**Comment by ASaiAnudeep** [comment]
Not sure, how to do it. We will look into it.

**Comment by leelaprasadv** [comment]
@rcurran1221

PactumJS mock server by default runs as http server since it will be a local mock server. 

Can you elaborate on your usecase that needs mock server to be on https ?

**Comment by 4ekki** [comment]
Hi. +1 for this request. Here's my usecase: 
I use Pactum as Mock Server, to test how my application interacts with dependencies. I start mock server, start my application, hit one of its endpoints with request and validate the response. On processing such request, my app sends a request to one of the dependencies, imposted by Pactum. It's real https request, not mocked on code level. 
And for some specific dependency we cannot use `http`, cause `https` is setup deeply in the code. So I want my mock server to be able to reply via https, not http. Therefore, I'm unable to use Pactum for this, cause it doesn't support working in `https` mode.

**Comment by leelaprasadv** [comment]
@4ekki would your code/setup work if pactum mock server runs on https with self signed certs on localhost?

**Reply by 4ekki** [reply]
@leelaprasadv, *looks like* it should work right out of the box, but I'm not 100% sure. Anyway though it will unblock me, cause I'll be able to test, ignoring https certificate errors (via NODE_TLS_REJECT_UNAUTHORIZED or smth). Now I'm facing some lower-level protocol errors and don't even get to the certificate details. 
But at some point it would be nice to have ability of using custom certificates as well. This will open a lot of scenarios for testing.

**Reply by 4ekki** [reply]
@leelaprasadv, @ASaiAnudeep, do you have any plans about implementing this?

**Reply by ASaiAnudeep** [reply]
@leelaprasadv is currently working on it.

**Comment by ASaiAnudeep** [comment]
Released in [v3.5.0](https://github.com/pactumjs/pactum/releases/tag/v3.5.0)

**Comment by madvi30477** [comment]
@ASaiAnudeep  and @leelaprasadv  can  Pactum Mock Server handles https requests and will it be able to handle client certificates with pfx format . i am using pactum 3.6.0 version but still facing issues while using mock server with https requests. and would like to know more about how to handle client certificates with pactumjs.

@4ekki @rcurran1221 : could you help me on how did u resolve this issue with pactumjs

**Reply by leelaprasadv** [reply]
@madvi30477 PactumJS currently supports https mock server using `.crt` and `.key`. We don't support `.pfx` at the moment.

In your case, if you do have the passphrase for your `.pfx`file, try converting it into .crt and key using openssl cli to use it. 

Hope this helps!

**Reply by madvi30477** [reply]
@leelaprasadv : Thanks for the prompt response, yes i have tried the same but i am getting invalid key every time. i have taken .pfx file and extracting key and certificate out f it, but i observed one thing during run time , pactum listening on http along with my https request . could u pls let me know in syntax of passing https url.
i am setting up the mock defaults as mentioned in pactumjs portal . 
eg const mockOpts = {port: 3001, host: '127.0.0.1', httpsOpts: {key: "server.key", cert: "server.crt"}};
await mock.setDefaults(mockOpts).

In place of host i have given https url but still i am getting invalid key where as same certificate and https url working fine in postman. could you pls help me to resolve the issue.

**Reply by madvi30477** [reply]
@leelaprasadv  : could you please help me on how to validate client certificates(.crt and .key files) using pactumjs. i have added  const mockOpts = {port: 3001, host: '127.0.0.1', httpsOpts: {key: "server.key", cert: "server.crt"}};
await mock.setDefaults(mockOpts). but its not validating client certificates during run time . Appreciate your response.

**Comment by madvi30477** [comment]
@leelaprasadv : could you please help me on how to validate mock service response while using mock.addInteractions.  Unable to understand mock server is running or not
